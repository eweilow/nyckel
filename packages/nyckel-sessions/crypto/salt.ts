export function saltKey(id: string, ...salt: string[]) {
  return [id, ...salt].join("");
}
