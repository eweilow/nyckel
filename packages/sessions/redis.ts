import { createClient, ClientOpts } from "redis";
import { EventEmitter } from "events";

export function createRedisClient(
  endpoint: string,
  handleError: (err: Error) => void,
  options?: ClientOpts
) {
  const client = createClient(endpoint, options);

  const healthyEventEmitter = new EventEmitter();

  let unhealthyTimeout: NodeJS.Timer | null = null;
  function markUnhealthy() {
    if (unhealthyTimeout != null) {
      return;
    }
    unhealthyTimeout = setTimeout(() => {
      healthy = false;
      healthyEventEmitter.emit("unhealthy");
      unhealthyTimeout = null;
    }, 5000);
  }

  let healthy: boolean = true;
  function markHealthy() {
    if (unhealthyTimeout != null) {
      clearTimeout(unhealthyTimeout);
      unhealthyTimeout = null;
    }

    if (!healthy) {
      healthy = true;
      healthyEventEmitter.emit("healthy");
    }
  }

  client.on("error", err => {
    handleError(err);
  });

  client.on("end", () => {
    markUnhealthy();
  });

  client.on("reconnecting", () => {
    markUnhealthy();
  });

  client.on("ready", () => {
    markHealthy();
  });

  markUnhealthy();

  return {
    client,
    get healthy() {
      return healthy;
    },
    healthyEventEmitter
  };
}
