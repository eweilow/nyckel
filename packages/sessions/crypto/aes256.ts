// encrypt and decrypt are adapted from https://github.com/JamesMGreene/node-aes256/blob/master/index.js
// Original license:
/*
The MIT License (MIT)

Copyright (c) 2015 James M. Greene

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import crypto from "crypto";

const rawTextEncoding = "utf8";
const encryptionEncoding = "base64";
const cipherType = "aes-256-ctr";

export function encrypt(key: string, plaintext: string) {
  if (typeof key !== "string" || !key) {
    throw new TypeError('Provided "key" must be a non-empty string');
  }
  if (typeof plaintext !== "string" || !plaintext) {
    throw new TypeError('Provided "plaintext" must be a non-empty string');
  }

  var sha256 = crypto.createHash("sha256");
  sha256.update(key);

  // Initialization Vector
  var iv = crypto.randomBytes(16);
  var cipher = crypto.createCipheriv(cipherType, sha256.digest(), iv);

  var ciphertext = cipher.update(Buffer.from(plaintext, rawTextEncoding));
  var encrypted = Buffer.concat([iv, ciphertext, cipher.final()]).toString(
    encryptionEncoding
  );

  return encrypted;
}

export function decrypt(key: string, encrypted: string) {
  if (typeof key !== "string" || !key) {
    throw new TypeError('Provided "key" must be a non-empty string');
  }
  if (typeof encrypted !== "string" || !encrypted) {
    throw new TypeError('Provided "encrypted" must be a non-empty string');
  }

  var sha256 = crypto.createHash("sha256");
  sha256.update(key);

  var input = Buffer.from(encrypted, "base64");

  if (input.length < 17) {
    throw new TypeError(
      'Provided "encrypted" must decrypt to a non-empty string'
    );
  }

  var iv = input.slice(0, 16);
  var decipher = crypto.createDecipheriv(cipherType, sha256.digest(), iv);

  var ciphertext = input.slice(16);
  var plaintext = Buffer.concat([
    decipher.update(ciphertext),
    decipher.final()
  ]).toString(rawTextEncoding);

  return plaintext;
}
