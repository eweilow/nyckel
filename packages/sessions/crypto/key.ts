import crypto from "crypto";
import { saltKey } from "./salt";

export function createKeyHash(id: string, salt: string): string {
  const cipher = crypto.createHash("sha256");
  cipher.update(saltKey(id, salt), "utf8");
  return cipher.digest("base64");
}
