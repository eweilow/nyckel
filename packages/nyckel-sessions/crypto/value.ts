import { saltKey } from "./salt";
import { encrypt, decrypt } from "./aes256";

function password(id: string, salt: string, secret: string) {
  return saltKey(id, salt, secret);
}

export function encryptData<T>(
  id: string,
  salt: string,
  secret: string,
  data: T
): string {
  const key = password(id, salt, secret);
  return encrypt(key, JSON.stringify({ data }));
}

export function decryptData<T>(
  id: string,
  salt: string,
  secret: string,
  encrypted: string
): T | null {
  try {
    const key = password(id, salt, secret);
    const decrypted = decrypt(key, encrypted);

    const { data } = JSON.parse(decrypted) as {
      data: T;
    };

    return data;
  } catch (err) {
    console.error(err);
    return null;
  }
}
