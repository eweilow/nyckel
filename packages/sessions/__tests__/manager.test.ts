import Redlock from "redlock";

import uuid from "uuid/v4";

jest.mock("../redis", () => {
  return {
    createRedisClient: jest.fn(() => ({
      client: {
        isThisARedisClient: true
      }
    }))
  };
});

jest.mock("redlock");

import { createSessionManager } from "..";
import { createRedisClient } from "../redis";
import { createKeyHash } from "../crypto/key";

describe("createSessionManager", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("has the correct signature", () => {
    expect(
      createSessionManager("url:redisEndpoint", () => {}, {
        salt: "option:salt",
        secret: "option:secret",
        ttl: 1000
      })
    ).toMatchSnapshot();
  });

  it("calls Redlock correctly", () => {
    const { client } = createSessionManager("url:redisEndpoint", () => {}, {
      salt: "option:salt",
      secret: "option:secret",
      ttl: 1000
    });
    expect((Redlock as any) as jest.Mock).toHaveBeenCalledWith([client], {
      retryCount: 0
    });
  });

  describe("properties", () => {
    describe(".client", () => {
      it("returns the client provided by Redis", () => {
        const { client: returnedClient } = createSessionManager(
          "url:redisEndpoint",
          () => {},
          {
            salt: "option:salt",
            secret: "option:secret",
            ttl: 1000
          }
        );
        const {
          client
        } = ((createRedisClient as any) as jest.Mock).mock.results[0].value;

        expect(client).toBe(returnedClient);
      });
    });

    describe(".isValidId(id)", () => {
      const { isValidId } = createSessionManager(
        "url:redisEndpoint",
        () => {},
        {
          salt: "option:salt",
          secret: "option:secret",
          ttl: 1000
        }
      );

      it("returns false for null or undefined id", () => {
        expect(isValidId(null as any)).toBe(false);
        expect(isValidId(undefined as any)).toBe(false);
      });

      it("returns false for non-string id", () => {
        expect(isValidId(3 as any)).toBe(false);
        expect(isValidId((() => {}) as any)).toBe(false);
        expect(isValidId(false as any)).toBe(false);
      });

      it("returns false for strings longer than 36", () => {
        expect(isValidId("x".repeat(37))).toBe(false);
        expect(isValidId("x".repeat(36))).toBe(true);
      });

      it("returns false for strings that are not alphanumeric and 36 chars long", () => {
        expect(isValidId("x".repeat(37))).toBe(false);
        expect(isValidId("x".repeat(36))).toBe(true);
        expect(isValidId("x".repeat(35))).toBe(false);

        expect(isValidId("åäö")).toBe(false);
      });

      it("returns true for UUID strings (random checks)", () => {
        const allowedTime = 500;
        let started = Date.now();

        while (Date.now() < started + allowedTime) {
          const string = uuid();
          expect({
            string,
            valid: isValidId(string)
          }).toEqual({
            string,
            valid: true
          });
        }
      });
    });
  });

  const salt = "option:salt";
  const keyToHash = (key: string) => createKeyHash(key, salt);
  function createManager() {
    return createSessionManager("url:redisEndpoint", () => {}, {
      salt,
      secret: "option:secret",
      ttl: 1000
    });
  }

  describe(".delete(id)", () => {
    function setupMock() {
      const client = {
        del: jest.fn((key, cb) => cb(null))
      };
      ((createRedisClient as any) as jest.Mock).mockReturnValueOnce({
        client
      });
      return client;
    }

    it("resolves if key exists", async () => {
      const client = setupMock();
      const manager = createManager();
      expect(await manager.delete("exists")).toBeUndefined();
      expect(client.del).toBeCalledTimes(1);
      expect(client.del).toBeCalledWith(
        keyToHash("exists"),
        expect.any(Function)
      );
    });

    it("resolves if key does not exist", async () => {
      const client = setupMock();
      const manager = createManager();
      expect(await manager.delete("non-existant")).toBeUndefined();
      expect(client.del).toBeCalledTimes(1);
      expect(client.del).toBeCalledWith(
        keyToHash("non-existant"),
        expect.any(Function)
      );
    });
  });

  describe(".has(id)", () => {
    function setupMock() {
      const client = {
        exists: jest.fn((key, cb) => cb(null, key === keyToHash("exists")))
      };
      ((createRedisClient as any) as jest.Mock).mockReturnValueOnce({
        client
      });
      return client;
    }

    it("resolves true if key exists", async () => {
      const client = setupMock();
      const manager = createManager();
      expect(await manager.has("exists")).toBe(true);
      expect(client.exists).toBeCalledTimes(1);
      expect(client.exists).toBeCalledWith(
        keyToHash("exists"),
        expect.any(Function)
      );
    });

    it("resolves false if key does not exist", async () => {
      const client = setupMock();
      const manager = createManager();
      expect(await manager.has("non-existant")).toBe(false);
      expect(client.exists).toBeCalledTimes(1);
      expect(client.exists).toBeCalledWith(
        keyToHash("non-existant"),
        expect.any(Function)
      );
    });
  });
});
