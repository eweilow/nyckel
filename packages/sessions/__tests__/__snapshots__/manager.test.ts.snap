// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSessionManager has the correct signature 1`] = `
Object {
  "delete": [Function],
  "generateId": [Function],
  "get": [Function],
  "has": [Function],
  "isValidId": [Function],
  "lock": [Function],
  "redisClient": Object {
    "client": Object {
      "isThisARedisClient": true,
    },
  },
  "set": [Function],
}
`;
